ELF support roadmap (my homework from the Linux training day) (#33317)
dotnet/runtime: docs/design/features/readytorun-elf-support-roadmap.md

Roadmap to native Linux executable format (ELF) support in Ready to Run
=======================================================================

Revisions:
* 1.1 - [Tomas Rylek](https://github.com/trylek) - 2/27/2020

# Introduction

This document details the steps needed for supporting native Linux executable

> format (ELF) for ready-to-run binaries on Linux. Today, ready-to-run binaries

Reviewer: Since ELF is supported by non Linux-based as well, could it be kept neutral from that perspective, e.g.:
```suggestion
This document details the steps needed for supporting native executable and linkable
format (ELF) for ready-to-run binaries on Unix-like operating systems. Today, ready-to-run binaries
```
or something that implies `systems that support ELF` (e.g. macOS uses `MACH-O`).

use the Windows PE executable format on all three supported platforms
(Windows, Linux, OSX). Due to this fact ready-to-run binaries are "unnatural"
on non-Windows platforms incurring both performance and user experience costs
(incomplete integration with OS-native application management, inability to

> use OS executable loader, lack of OS-native diagnostic experience). This

Reviewer 2: There are different gradiations of 'OS-native' diagnostics. I am guessing we are proposing only a very scoped set of improvements at this point? An example progression given increasing levels of effort:
1. Use ELF format - tools such as readelf and basic module enumeration in debuggers work
2. Add ELF dynamic symbol table - debuggers/profilers should be able to resolve IPs in those methods to names
3. Add DWARF information for all methods + line information - debuggers/profilers can resolve all pre-compiled methods to names and source, jitted code is still opaque
4. .NET type information encoded with DWARF symbols - Limited object inspection works. Functionality that isn't C-like still probably doesn't work such as variable length types (string/array) or generic instantiations not known at image compile time. 
5. .Net specific DWARF extensions + 1st class support for .Net built into standard tools - Everything works : )

Reviewer 3: It's a shame that the GCC LTO information isn't stable and depends heavily on GIMPLE (GCC's IL); otherwise Crossgen2 could emit it and we could get LTO "for free". (Don't know about Clang's approach here, but I would imagine it depends heavily on the LLVM IR.)

Reviewer 4: > GCC LTO information

A few problems with it:
- The intermediate representation of C/C++ compilers is not aware of GC and it is missing other constructs required to support managed code well. Same problems exist with clang.
- GCC has incompatible license

  the composite R2R effort; this will replace the existing lookup for COR header
  using the COM descriptor directory entry in the PE header.

* **Crossgen2 compiler**: The SectionBuilder and R2RPEBuilder needs to be
  decoupled from the PE format by some form of virtualization. We need to
  port the various ELF header structures to managed code for use in Crossgen2
  and R2RDump. Once we have the constants, structures and their serialization
  mechanisms in place and after we figure out the proper abstraction the work
  should be pretty straightforward in terms of just filling in the various header
  and section tables and emitting the section data in the proper place.

* **Diagnostics / SOS**: switching over to native ELF will have fundamental effect on
  diagnostics. While the overall long-term effect is assumed to be positive,
  it will take time to bridge the user experience gap. It is likely that for some
  temporary period of time we'll need to keep support for both PE files and ELF
  files on Linux until all diagnostic functionality for ELF has been put on par
  with pre-existing diagnostic support for PE-based executables.

> * **Visual Studio, VS Code**: switching over the executable format is certain to

Reviewer 2: Windbg is also important here if we plan to use these images in Azure. Windbg gets used internally in Azure Watson to debug Linux dumps and we are working to improve its ability to handle managed code.

  have non-trivial effect on both VS and VS Code. We need to plan ahead with the
  relevant teams to make sure we cover all the relevant project template, build
  system, debugger, profiler and other aspects.

* **R2RDump**: The PE header reader that is today using the managed assembly
  System.Reflection.PortableExecutable needs to be generalized to recognize and
  parse the ELF headers and locate the RTR_HEADER export symbol - once we're able
  to identify the R2R header in the file, we immediately have access to all the
  sections and data. Naturally we'll also need to adjust the current logic for
  translating RVA's to file offsets.

* **ILSpy**: will hopefully continually ingest R2RDump changes so it should
  automatically pick up support for ELF-embedded MSIL/ECMA-335 metadata.

# Proposed sequencing

The natural sequencing and initial T-shirt costing implied by the various
inter-dependencies and by the natural requirement that each step needs to be
verifiable / diagnosable is as follows:

* Import (rewrite) the relevant ELF headers to managed code, probably

>   somewhere under *src/coreclr/src/tools/Common/Internal/Runtime*.

Reviewer 4: This can use backticks as well.

Author: Fixed in 5th commit.

  This is junior level work estimated at 2 man-weeks including some rudimentary
  regression testing - e.g. writing a simple test that will dump a subset of
  the headers for a hard-coded ELF file used as a resource by the test.
  
  An interesting observation is that most ELF header structures have
  parallel 32-bit and 64-bit variants. It might be useful to logically unify them
  by internally using longs and just have virtual de-serializers that would have
  something like ReadNativeInt representing 32-bit or 64-bit integer based on
  the bitness.
  
  We probably don't need to support ELF to the fullest, e.g. endianness doesn't

>   matter as our JIT and runtime doesn't support any high-endian CPU's (ARM has

Reviewer 3: ```suggestion
  matter as our JIT and runtime doesn't support any big-endian CPU's (ARM has
```

  technically switchable endianness but we only support the little variant).

* Create a simplified counterpart of System.Reflection.PortableExecutable to
  support ELF, again somewhere under ``src/coreclr/src/tools/Common/Internal/Runtime``.
  At some point we may consider porting it to libraries but initially it's better
  to keep that separate.
  

>   We should also double-check whether mono doesn't already have something in this

Reviewer 5: Haven't looked at this in detail, but this came up on my Twitter feed recently: https://github.com/xoofx/LibObjectFile. It also comes with some level of DWARF support.

  vein we could reuse or leverage. We basically need to be able to read / write

>   the main ELF header, the program headers, the section headers and the symbols.

Reviewer 6: Eventually, the ELF R2R assemblies will be loaded via dlopen() on Linux, right? 
It is still necessary to be able to load the assemblies by mapping the individual sections in order to support single-exe apps.

  the choice of PE / ELF format and use the newly created ELF API to [re]construct
  the necessary headers and create a fully functional ELF executable.

  At this point this will probably still be just a dynamic library, not a
  full-fledged EXE. The full-fledged EXE should come in parallel with the
  single-exe effort as it basically involves emitting some host bootstrapper into
  the executable. Without the bootstrapper we can happily continue doing more or
  less what we're doing today just with slightly different logic for locating
  the headers.

* Implementing CoreCLR runtime counterpart of PEFile and PEImage for ELF and
  adding Linux-specific format detection to choose between PE and ELF as it's
  expected that for a certain period of time we'll need to support both formats
  on Linux. This is a senior / principal-level task estimated at 4 man-weeks
  of work.
  
  It will require initial investigation as to what native code we can reuse
  from the existing Linux libraries. We should consider some level of
  virtualization to cater for common constructs. The new objects should be
  Linux-specific and should use the Linux-native functions for manipulating

>   executables like **dlopen** / **dlsym**.

Reviewer 4: Nit: The markdown convention is to quote identifiers using backticks

(multiple places)

Author: Fixed even though this is actually the only place I found - are you aware of others?

use the Windows PE executable format on all three supported platforms
(Windows, Linux, OSX). Due to this fact ready-to-run binaries are "unnatural"
on non-Windows platforms incurring both performance and user experience costs
(incomplete integration with OS-native application management, inability to
use OS executable loader, lack of OS-native diagnostic experience). This
proposal tries to summarize the steps needed towards native executable support.

# Scope of changes

Support for Linux native executable format will require changes to the following
areas:

* **CoreCLR runtime**: ideally we should be able to use the Linux ELF headers
  in the native code. The current treatment of ready-to-run files involving
  classes PEFile, PEImage and PEImageLayout aka PEDecoder, need virtualizing
  to allow for their ELF counterparts and proper deduplication of common logic
  especially related to R2R that should be mostly autonomous once the R2R header
  has been located. Fully switching over to native ELF format for ready-to-run
  binaries is more complicated as it involves not yet implemented embedded MSIL
  in native R2R binaries, a new concept we're currently developing as part of

>   the composite R2R effort; this will replace the existing lookup for COR header

Reviewer 4: There is way too many places in the ecosystem that depend on the IL .dll files on disk. I do not think it will replace it. It may be one more option.

Author: Are you suggesting any specific change to the wording here?

Reviewer 4: I would delete the whole sentence "this will replace the existing lookup for COR header using the COM descriptor directory entry in the PE header"

  using the COM descriptor directory entry in the PE header.

* **Crossgen2 compiler**: The SectionBuilder and R2RPEBuilder needs to be
  decoupled from the PE format by some form of virtualization. We need to
  port the various ELF header structures to managed code for use in Crossgen2
  and R2RDump (do we need some approval or special arrangements for porting

>   the Linux header files like elf.h and elftypes.h to managed code in the runtime

Reviewer 4: This is documented in https://github.com/dotnet/runtime/blob/master/CONTRIBUTING.md#copying-files-from-other-projects

We have been using these headers from LLVM that have compatible license: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/inc/llvm/ELF.h

Reviewer 4: So do not talk about these as "Linux header files" because they are unlikely to come from Linux due to incompatible license.

Author: Fixed in 3rd commit, thanks for clarifying!

  repo?). After we have the constants, structures and their serialization
  mechanisms in place and after we figure out the proper abstraction the work
  should be pretty straightforward in terms of just filling in the various header
  and section tables and emitting the section data in the proper place.

> * **Diagnostics / SOS**: switching over to native ELF will have fundamental effect on

Reviewer 4: I do not see why switching to ELF envelope for R2R has fundamental effect on SOS

Author: OK, JanV also mentioned offline some of these might be no-concerns.

  diagnostics. While the overall long-term effect is assumed to be positive,
  it will take time to bridge the user experience gap. It is likely that for some
  temporary period of time we'll need to keep support for both PE files and ELF
  files on Linux until all diagnostic functionality for ELF has been put on par

>   with pre-existing diagnostic support for PE-based executables.

Reviewer 7: The biggest place I see impact on the diagnostic tools is on profiling, and possibly native debugging. Right now we use crossgen to generate a native pdb to interface with the profiler, but with the generation of ELF files directly, I believe we would also want to at least cost out the effort of generating DWARF debugging information for these files. Providing a means for pushing some degree of managed debugging into the native Unix debugging stack would be quite valuable, and potentially reduce the problems caused by a lack of support for mixed mode debugging in managed code. But, as Jan notes, these aren't concerns about modifying existing tools so much as enabling entire new categories of capability.

Author: Thanks David for pointing that out, I believe this should be definitely somehow made part of this design doc even though the actual prioritization is yet to be determined.

Reviewer 2: > cost out the effort of generating DWARF debugging information

For debugging it could make some difference, profiling I'm not sure it would. On Linux we generate perfmaps for the 'perf' tool already and that I'm not sure if we lack anything that DWARF would give us instead. For Visual Studio profiler resolving R2R image symbols doesn't currently work - we are discussing getting the symbolic data into the trace at rundown rather than out-of-line tools to create symbol formats.

  parse the ELF headers and locate the RTR_HEADER export symbol - once we're able
  to identify the R2R header in the file, we immediately have access to all the
  sections and data. Naturally we'll also need to adjust the current logic for
  translating RVA's to file offsets.

* **ILSpy**: will hopefully continually ingest R2RDump changes so it should
  automatically pick up support for ELF-embedded MSIL/ECMA-335 metadata.

# Proposed sequencing

The natural sequencing and initial T-shirt costing implied by the various
inter-dependencies and by the natural requirement that each step needs to be
verifiable / diagnosable is as follows:

* Import (rewrite) the relevant Linux headers elf.h and elftypes.h to managed
  code, probably somewhere under *src/coreclr/src/tools/Common/Internal/Runtime*.
  We need to verify licensing conditions for doing this. The actual work is
  junior level, estimated at 2 man-weeks including some rudimentary regression
  testing - e.g. writing a simple test that will dump a subset of the headers
  for a hard-coded ELF file used as a resource by the test.
  
  An interesting observation is that most ELF header structures have

>   paraller 32-bit and 64-bit variants. It might be useful to logically unify them

Reviewer 4: ```suggestion
  parallel 32-bit and 64-bit variants. It might be useful to logically unify them
```

Author: Thanks, I actually noticed another typo, will fix both immediately

Author: OK, I cannot find "the other typo" right now. I fixed this one in the 2nd commit.

  section builder logic in a similar manner for both PE and ELF builder - this
  would probably require creating a wrapper over the PE builder but it would be
  still easier than having two completely different code paths for composing the
  sections. We'll need to address minute differences like alignment considerations
  as on Linux **file offset - RVA** deltas need to be page-aligned.

* Virtualizing R2RDump header parsing to add ELF support. This should be
  junior-level task estimated at 1 man-week of work. It should check for the
  well-known signature magic values used by PE / ELF files to identify the format
  of the input file (both for the "main" and reference input files) and use
  proper logic to locate the R2R header, COR header / MetadatReader and the logic
  to map RVA's to file offsets.

* Virtualizing Crossgen2 SectionBuilder / R2RPEBuilder: this is senior / principal
  work estimated at 3 man-weeks to refactor section building to be independent on
  the choice of PE / ELF format and use the newly created ELF API to [re]construct
  the necessary headers and create a fully functional ELF executable.

>   At this point this will probably still be just a dynamic library, not a

Reviewer 4: For this, crossgen needs to emit `.o` file and the final binary needs to be produced by platform linker (ie `ld` on Linux).

Author: OK, I guess this means we basically need to be able to generate an object or executable flavor of the ELF file. At least I hope they differ less than obj and exe on Windows :-).

Reviewer 6: Generating `.o` from crossgen and linking later would be very useful feature -- this will enable building native linked single-exe apps. 

However, how will cross compilation of apps work? We'll need cross-platform native tool chains (ex: `ld` on Windows)
